### ISSUE TITLE
Reinvestigate command-line input of associative arrays

### ISSUE THREAD

**Issue by schlessera** [wp-cli/wp-cli#4616]
WP-CLI includes some arguments that need an associative array as input (like `wp post create --meta_input`), but it does not support any way of entering these on the command line.

A previous issue on this was closed in favor of handling CRUD operations through the WP REST API, however this is far from reliable/complete and we will not switch to using that method as a new default anytime soon.

This is why I want to reevaluate how we could process shell input to accept associative (and maybe even nested) arrays.

My current thought is to have a utility function to pass all these arguments through, like `WP_CLI\Utils\parse_array()`, which would try to parse the provided input as either JSON array syntax (`["foo":"bar"]`) or PHP array syntax (`["foo"=>"bar"]`). Whatever arguments need (associative/nested) arrays would run their value through this function, and in case it returns false, through an error that the argument could not be parsed/is not a valid array.

We'll probably detect lots of edge cases over time. Having this parsing in one central location makes bug fixes across all commands easy.

Related: https://github.com/wp-cli/wp-cli/issues/1323

**Comment by johnbillion** [comment]
FWIW I would favour JSON syntax over PHP syntax, as it's more portable and more straight forward to parse.

**Comment by schlessera** [comment]
Yes, I agree, JSON is probably easier. I'd like to test whether we can make both work, so people can copy-paste array renderings in there. But at least JSON should be functional.

**Comment by gitlost** [comment]
Removing this from the 1.5.0 milestone in favor of https://github.com/wp-cli/wp-cli/issues/4638

**Comment by kirtangajjar** [comment]
@schlessera @gitlost Looks like we can close this as we have `parse_shell_arrays`.

**Comment by schlessera** [comment]
@kirtangajjar I had left it open because I'd like to go through the commands and make a list of parameters that would require/benefit the use of `parse_shell_arrays()`. So far, we have only implemented this for 1 single parameter, I'm sure there are more that should be converted.

**Comment by TimothyBJacobs** [comment]
What about using array dot syntax? I think that is easier to write by hand, particularly when dealing with quoting.

For instance:

```
--content.raw="Hello O'Leary"
--content='{ "raw": "Hello O\'Leary" }'
```

That would also make it possible to define the parameters in the synopsis.

```
--content.raw=<content.raw>
: The raw content for the post
```

In my mind, WP CLI would transform the dotted syntax and pass a `content` assoc arg as an array with the `raw` property. I imagine the base property could also accept JSON syntax.

I don't think there would BC issues since AFAICT, dotted parameter names are currently rejected by the synopsis validator.

**Comment by TheCrowned** [comment]
Any update on this? There are multiple scenarios where an array is supported (and the only way to actually do something), but WP CLI won't support it. In my case, it's `wp comment list --date_query`. I'd be willing to maybe work on this myself and submit a PR, if we get a clear idea of the direction :)

**Comment by usizu** [comment]
Is there any way to input an array to a WP CLI command? For example `wp wc product update --attributes=[<array here>]`

**Comment by DrLightman** [comment]
Still no support for arrays for wp-cli? date_query, tax_query, etc

**Comment by swissspidy** [comment]
Chiming in here because of https://github.com/wp-cli/entity-command/issues/519

> I had left it open because I'd like to go through the commands and make a list of parameters that would require/benefit the use of parse_shell_arrays()

Probably better to just do add this on a case-by-case basis in each repo, after all this time.

> What about using array dot syntax? I think that is easier to write by hand, particularly when dealing with quoting.

Interesting, but by now the JSON method is already established elsewhere and I don't see us introducing a second way of doing things here.

> Any update on this? There are multiple scenarios where an array is supported (and the only way to actually do something), but WP CLI won't support it. In my case, it's `wp comment list --date_query`. I'd be willing to maybe work on this myself and submit a PR, if we get a clear idea of the direction :)

We have `\WP_CLI\Utils\parse_shell_arrays` that we can use in places like this, so this shouldn't be blocked. See https://github.com/wp-cli/entity-command/pull/133 for an example.

> Is there any way to input an array to a WP CLI command? For example `wp wc product update --attributes=[<array here>]`

For custom commands the developer will need to add support, in this case I'd bring this up with WooCommerce.

> Still no support for arrays for wp-cli? date_query, tax_query, etc

https://github.com/wp-cli/entity-command/issues/519 was just opened to request this. If someone wants to contribute that enhancement, feel free :)
